### DISCUSSION TITLE
Returning fields computed/derived from values of other fields

### DISCUSSION THREAD

**Discussion by Norbo11** [fastapi/fastapi#8707] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class MyModel(BaseModel):

    a: str
    b: str

    @property
    def c():
        return a + b
```

### Description

I'm trying to make FastAPI return some extra fields as part of my JSON response which is based on the values of some other fields.

I can't find any mention of this in the docs.

I don't want to add the derived field and mutate the object whenever its dependencies update, as that seems a lot more error prone. Ideally I'd love to do something like in the example code snippet, where `c` is defined to always be `a + b` and this gets evaluated whenever a response is serialised.

Thanks!

### Operating System

Linux

### FastAPI Version

0.63.0

### Python Version

3.8.7

**Comment by insomnes** [comment]
You can find some approaches here in corresponding [pydantic issue](https://github.com/samuelcolvin/pydantic/issues/935).

**Comment by kaiix** [comment]
this is an issue of pydantic. I'll suggest you use `validator` for a computed property

```py
from typing import Optional

from pydantic import BaseModel, validator


class MyModel(BaseModel):

    a: str
    b: str
    c: Optional[str]

    @validator('c', pre=True, always=True)
    def make_c(cls, v: str, values: dict):
        return values['a'] + values['b']
```

ref: https://pydantic-docs.helpmanual.io/usage/validators/#validate-always

**Comment by dylanwa** [comment]
@kaiix , I think this only works when you submit a request with field 'c', the validator won't be hit if the request only has 'a' and 'b', I think what @Norbo11 want is
"user submit request of 'a' and 'b', it will hit endpoint with 'a', 'b' and 'c' (c=a+b)"

--edit--
sorry I missed the feature "validator always" with parameter "always=True", it do the trick :)

**Comment by Norbo11** [comment]
@dylanwa indeed, I'm using root_validator and it's working quite well:

```python
from abc import abstractmethod

from pydantic import BaseModel, root_validator


class BaseModelWithStringRepresentation(BaseModel):

    string_representation: str = ""

    @classmethod
    @abstractmethod
    def get_string_representation(cls, values: dict):
        pass

    @root_validator
    def validate_string_representation(cls, values: dict):
        values["string_representation"] = cls.get_string_representation(values)
        return values
```

I have models subclassing `BaseModelWithStringRepresentation` and implementing `get_string_representation` where they will return a value based on some computation on their fields. Thanks @kaiix for the suggestion.

**Comment by ericman93** [comment]
what is the best solution for adding calculated fields from the DB?
I have output model and it's `dict` method I added fetches to return to the user
but `dict` is sync method and my db queries can be async

**Comment by dynajosh** [comment]
Hi eric, did you find a solution for this?

**Comment by mdhvcd** [comment]
> @kaiix , I think this only works when you submit a request with field 'c', the validator won't be hit if the request only has 'a' and 'b', I think what @Norbo11 want is "user submit request of 'a' and 'b', it will hit endpoint with 'a', 'b' and 'c' (c=a+b)"
> 
> --edit-- sorry I missed the feature "validator always" with parameter "always=True", it do the trick :)

what if I want that c value should be calculated after query hit from the database like in serializer(django) we just over ride get_c function and get the value according to our need

**Comment by ericman93** [comment]
@dynajosh not really, I created an instance of an output model in the controller when the method is async

**Comment by ericman93** [comment, marked as answer]
@dynajosh I change the solution to this one
I created my own base class with a `to_output_model` method that returns an output instance with the calculated field (by default returns self)

```
class MyBase(BaseModel):
   async def to_output_model(self):
       return self

class Credit(MyBase):
  current: int
  total: int

  async def to_output_model(self):
      return OutputCredit(
          current=self.current,
          total=self.total,
          remaining=self.total - self.current
      )

class OutputCredit(Credit):
  remaining: int
```

but when adding a method to the app, I'm adding a wrapper method (in my case I created a decorator for it)
and the wrapper method returns the output model instead the db model

```
def create_wrapped_func(func: Callable):
    async def wrapper_func(*args, **kwargs) -> Any:
        # you can support sync methods as well
        result = await func(*args, **kwargs)
        return await _to_output_model(result)

    return wrapper_func

async def _to_output_model(result: Any):
    # todo: support dicts as well
    if isinstance(result, MyBase):
        return await result.to_output_model()
    elif isinstance(result, list):
        return [await _to_output_model(item) for item in result]
    else:
        return result

# you can create a wapper of your own for this one
app.post("/")(create_wrapped_func(my_method))
```

the only part that I'm not sure about, is that the DB model is familiar with the output model
